### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by aknysh**
## what
* Terraform module to provision an `Elasticsearch` cluster in a VPC with build-in `Kibana` and `Logstash`

## why
* Easily provision `Elasticsearch` cluster with the specified node count in the provided subnets in a VPC
* Provision Elasticsearch domain policy that accepts a list of IAM role ARNs from which to permit management traffic to the cluster
* Provision Security Group to control access to the Elasticsearch domain (inputs to the Security Group are other Security Groups or CIDRs blocks to be allowed to connect to the cluster)
* Provision DNS hostname record for Elasticsearch cluster
* DNS hostname record for Kibana

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  log_publishing_options {
+    enabled                  = "${var.log_publishing_enabled }"
+    log_type                 = "${var.log_publishing_log_type}"
+    cloudwatch_log_group_arn = "${var.log_publishing_cloudwatch_log_group_arn}"
+  }
+
+  tags = "${module.label.tags}"
+}
+
+resource "aws_elasticsearch_domain_policy" "es_vpc_management_access" {
```
Comment (osterman) [inline main.tf]: from `es_` prefix as we do not need the disambiguation.

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+resource "aws_security_group" "default" {
+  count       = "${var.enabled == "true" ? 1 : 0}"
+  vpc_id      = "${var.vpc_id}"
+  name        = "${module.label.id}"
+  description = "Allow inbound traffic from Security Groups and CIDRs. Allow all outbound traffic"
+  tags        = "${module.label.tags}"
+}
+
+resource "aws_security_group_rule" "ingress_security_groups" {
```
[the remark this answers is not in the document]
Comment (aknysh) [reply main.tf]: good pint.
they are separate resources and having different roles to access them would be preferable.

File: README.md [hunk 3]
```diff
@@ -1 +1,276 @@
...
+- [VPC Support for Amazon Elasticsearch Service Domains](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-vpc.html) - Describes Elasticsearch Service VPC Support and VPC architectures with and without zone awareness
+- [Creating and Configuring Amazon Elasticsearch Service Domains](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-createupdatedomains.html) - Provides a complete description on how to create and configure Amazon Elasticsearch Service (Amazon ES) domains
+- [Kibana and Logstash](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-kibana.html) - Describes some considerations for using Kibana and Logstash with Amazon Elasticsearch Service
+- [Amazon Cognito Authentication for Kibana](https://docs.aws.amazon.com/elasticsearch-service/latest/developerguide/es-cognito-auth.html) - Amazon Elasticsearch Service uses Amazon Cognito to offer user name and password protection for Kibana
+- [elasticsearch_domain](https://www.terraform.io/docs/providers/aws/r/elasticsearch_domain.html) - Configuration for `elasticsearch_domain` Terraform resource
```
Comment (osterman) [inline README.md]: Terraform reference documentation for the `elasticsearch_domain` resource

File: README.md [hunk 4]
```diff
@@ -1 +1,276 @@
...
+- [elasticsearch_domain_policy](https://www.terraform.io/docs/providers/aws/r/elasticsearch_domain_policy.html) - Configuration for `elasticsearch_domain_policy` Terraform resource
```
Comment (osterman) [inline README.md]: Terraform reference documentation for the `elasticsearch_domain_policy` resource

File: main.tf (hunk 2 above)
[the remark this answers is not in the document]
Comment (aknysh) [reply main.tf]: let's move the decision of `who can access what` out of the module.
Top-level modules will provide:

```hcl
variable "iam_role_arns" {
  type        = "list"
  default     = []
  description = "List of IAM role ARNs to permit access to the Elasticsearch domain"
}

variable "iam_actions" {
  type        = "list"
  default     = []
  description = "List of actions to allow for the IAM roles, _e.g._ `es:ESHttpGet`, `es:ESHttpPut`, `es:ESHttpPost`"
}
```

**Comment by osterman** [comment]
Travis is failing

**Pull request merged**